rmw_cyclonedds_cpp/src/rmw_node.cpp

```diff
@@ -861,107 +859,79 @@ rmw_ret_t configure_qos_for_security(
-    return RMW_RET_INCORRECT_RMW_IMPLEMENTATION);
-
-  /* domain_id = UINT32_MAX = Cyclone DDS' "default domain id".*/
-  if (options->domain_id >= UINT32_MAX) {
-    RCUTILS_LOG_ERROR_NAMED(
-      "rmw_cyclonedds_cpp", "rmw_create_node: domain id out of range");
-    return RMW_RET_INVALID_ARGUMENT;
-  }
-  const dds_domainid_t domain_id = static_cast<dds_domainid_t>(options->domain_id);
-
-  context->instance_id = options->instance_id;
-  context->implementation_identifier = eclipse_cyclonedds_identifier;
-  context->impl = nullptr;
-
-  if ((ret = rmw_init_options_copy(options, &context->options)) != RMW_RET_OK) {
-    return ret;
-  }
-
-  std::unique_ptr<rmw_context_impl_t> impl(new(std::nothrow) rmw_context_impl_t());
-  if (impl == nullptr) {
-    return RMW_RET_BAD_ALLOC;
+  std::lock_guard<std::mutex> guard(initialization_mutex);
+  if (1u != ++this->node_count) {
+    // initialization has already been done
+    return RMW_RET_OK;
   }
 
   /* Take domains_lock and hold it until after the participant creation succeeded or
-     failed: otherwise there is a race with rmw_destroy_node deleting the last participant
-     and tearing down the domain for versions of Cyclone that implement the original
-     version of dds_create_domain that doesn't return a handle.  */
-  if (!check_create_domain(domain_id, options->localhost_only)) {
+    failed: otherwise there is a race with rmw_destroy_node deleting the last participant
+    and tearing down the domain for versions of Cyclone that implement the original
+    version of dds_create_domain that doesn't return a handle.  */
+  this->domain_id = static_cast<dds_domainid_t>(options->domain_id);
+  if (!check_create_domain(this->domain_id, options->localhost_only)) {
     return RMW_RET_ERROR;
   }
 
-  /* Once the domain id is set in impl, impl's destructor will take care of unref'ing
-     the domain */
-  impl->domain_id = domain_id;
-
   std::unique_ptr<dds_qos_t, std::function<void(dds_qos_t *)>>
   ppant_qos(dds_create_qos(), &dds_delete_qos);
   if (ppant_qos == nullptr) {
     return RMW_RET_BAD_ALLOC;
   }
   std::string user_data = std::string("enclave=") + std::string(
-    context->options.enclave) + std::string(";");
+    options->enclave) + std::string(";");
   dds_qset_userdata(ppant_qos.get(), user_data.c_str(), user_data.size());
   if (configure_qos_for_security(
       ppant_qos.get(),
-      &context->options.security_options) != RMW_RET_OK)
+      &options->security_options) != RMW_RET_OK)
   {
-    if (context->options.security_options.enforce_security == RMW_SECURITY_ENFORCEMENT_ENFORCE) {
+    if (RMW_SECURITY_ENFORCEMENT_ENFORCE == options->security_options.enforce_security) {
       return RMW_RET_ERROR;
     }
   }
-  impl->ppant = dds_create_participant(domain_id, ppant_qos.get(), nullptr);
-  if (impl->ppant < 0) {
+
+  this->ppant = dds_create_participant(this->domain_id, ppant_qos.get(), nullptr);
+  if (this->ppant < 0) {
     RCUTILS_LOG_ERROR_NAMED(
```

Comment:
If `dds_create_participant()` fails, is it guaranteed that it internally "unref'd" the domain? I doesn't look like `~rmw_context_impl_t()` can be relied on as before because there's no attempt to clean up the domain in there.


https://github.com/ros2/rmw_cyclonedds/blob/3b9bd8056de24c7e5c0ff06c44c3dca6adc4bb11/rmw_cyclonedds_cpp/src/rmw_node.cpp#L280-L289

Reply:
When `dds_create_participant` fails, `clean_up` is called which calls `check_destroy_domain`, which is "unref'd" the domain.

Reply:
What if the `dds_create_reader()` calls below fail? Do those unref the participant and domain? It looks like they might also return from this function without cleaning up.

Reply:
This ones? https://github.com/ros2/rmw_cyclonedds/blob/ded89f9ab699905649468ff68e128f3587f9c09d/rmw_cyclonedds_cpp/src/rmw_node.cpp#L926-L949

All of them also call `clean_up` on failure.
I agree that RAII would be nicer here, but I think this is cleaning up things correctly.